src/bits.rs

```diff
@@ -0,0 +1,429 @@
+                }
+
+                #[inline(always)]
+                fn wrapping_shr(self, rhs: usize) -> Self {
+                    <$T>::wrapping_shr(self, rhs as u32)
+                }
+            }
+        )+
+    }
+}
+
+impl_int_bits! {
+    i8 i16 i32 isize
+    u8 u16 u32 usize
+}
+
+#[cfg(any(target_arch = "x86_64"))]
+impl_int_bits! {
+    i64 u64
+}
+
+impl_int_bits! {
+    i128 u128
+}
+
+/// A binary representation of an array.
+pub trait BitsArray<T: Bits>: Array<Item = T> {
+    /// Returns the number of ones in the binary representation of `self`.
+    #[inline]
+    fn count_ones(&self) -> usize {
+        self.as_slice()
+            .iter()
+            .fold(0, |acc, bits| acc + bits.count_ones())
+    }
+
+    /// Returns the number of zeros in the binary representation of `self`.
+    #[inline]
+    fn count_zeros(&self) -> usize {
+        self.as_slice()
+            .iter()
+            .fold(0, |acc, bits| acc + bits.count_zeros())
+    }
+
+    /// Returns the number of leading zeros in the binary representation of `self`.
+    #[inline]
+    fn leading_zeros(&self) -> usize {
+        let mut acc = 0;
+
+        for bits in self.as_slice().iter() {
+            if *bits == T::ZERO {
+                acc += T::BITS;
+            } else {
+                acc += bits.leading_zeros();
+                break;
+            }
+        }
+
+        acc
+    }
+
+    /// Returns the number of trailing zeros in the binary representation of `self`.
+    #[inline]
+    fn trailing_zeros(&self) -> usize {
+        let mut acc = 0;
+
+        for bits in self.as_slice().iter().rev() {
+            if *bits == T::ZERO {
+                acc += T::BITS;
+            } else {
+                acc += bits.trailing_zeros();
+                break;
+            }
+        }
+
+        acc
+    }
+
+    /// Clears the bits.
+    #[inline]
+    fn clear(&mut self) {
+        for i in self.as_mut_slice().iter_mut() {
+            *i = T::ZERO;
+        }
+    }
+
+    /// Returns the boolean representation of a bit.
+    #[inline]
+    fn get_bit(&self, index: usize) -> Option<bool> {
+        if index >= self.len() * T::BITS {
+            None
+        } else {
+            let word = index >> T::INDX;
+            let bit = index & T::MASK;
+
+            Some(self.as_slice()[word] >> bit & T::ONE == T::ONE)
+        }
+    }
+
+    /// Sets the value of a bit.
+    #[inline]
+    fn set_bit(&mut self, index: usize, val: bool) {
+        if index < self.len() * T::BITS {
+            let word = index >> T::INDX;
+            let bit = index & T::MASK;
+
+            if val {
+                self.as_mut_slice()[word] |= T::ONE << bit;
+            } else {
+                self.as_mut_slice()[word] &= !(T::ONE << bit);
+            }
+        }
+    }
+
+    /// Returns the offset of the next `1` bit that follows `after`.
+    #[inline]
+    fn find_next_one(&self, after: Option<usize>) -> Option<usize> {
```

Comment:
This is probably better done using an iterator. Then you can keep a word in the iterator which you modify in each step by removing the lowest bit by doing `x & (x - 1)`. Advance to the next word when the current word is zero.